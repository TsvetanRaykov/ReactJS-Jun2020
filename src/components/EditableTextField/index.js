import React, { useState, useEffect, createRef } from 'react'
import { Tooltip, withStyles, TextField } from '@material-ui/core'

const styles = (theme) => ({
	userName: {
		fontSize: '1rem',
	},
})

const EditableTextField = (props) => {
	const { text, callback, classes } = props

	const [isInEditMode, setEdiMode] = useState(false)
	const [newText, setNewtext] = useState(text)

	const textInput = createRef()

	const chageToEditMode = () => {
		setEdiMode(true)
	}

	const saveText = () => {
		if (callback && newText !== text) {
			callback(newText)
		}
		setEdiMode(false)
	}

	useEffect(() => {
		setNewtext(text)
	}, [text])

	useEffect(() => {
		if (textInput.current) {
			textInput.current.focus()
		}
	}, [textInput])

	const renderEditView = () => {
		return (
			<div>
				<TextField
					onChange={(e) => setNewtext(e.target.value)}
					onBlur={saveText}
					value={newText}
					ref={textInput}
					autoFocus
				/>
			</div>
		)
	}

	const renderDefaultView = () => (
		<Tooltip placement='bottom-start' title='Double click to edit'>
			<div className={classes.userName} onDoubleClick={chageToEditMode}>
				{newText}
			</div>
		</Tooltip>
	)

	return isInEditMode ? renderEditView() : renderDefaultView()
}

export default withStyles(styles)(EditableTextField)
