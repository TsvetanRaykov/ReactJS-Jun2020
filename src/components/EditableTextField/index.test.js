import React from 'react'
import renderer from 'react-test-renderer'
import TestWrapper from '../../utils/test-wrapper'
import EditableTextField from '.'

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useLayoutEffect: jest.requireActual('react').useEffect,
}))

describe('<EditableTextField />', () => {
	it('should match the snapshot', () => {
		const wrapper = renderer.create(
			<TestWrapper>
				<EditableTextField />)
			</TestWrapper>
		)
		expect(wrapper).toMatchSnapshot()
	})
})
