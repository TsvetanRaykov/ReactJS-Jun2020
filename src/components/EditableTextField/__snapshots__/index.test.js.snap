// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableTextField /> should match the snapshot 1`] = `
Array [
  <div
    aria-describedby={null}
    className="EditableTextField-userName-1"
    onBlur={[Function]}
    onDoubleClick={[Function]}
    onFocus={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    title="Double click to edit"
  />,
  ")",
]
`;
